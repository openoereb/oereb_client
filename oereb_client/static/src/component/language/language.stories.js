import React from "react";

import OerebLanguage from "./language";

export default {
  title: 'API Reference/Component/Language',
  component: OerebLanguage
};

export const Language = () => <OerebLanguage />;
