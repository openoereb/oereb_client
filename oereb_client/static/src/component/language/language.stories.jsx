import React from "react";

import OerebLanguage from "./language";

export default {
  title: "API Reference/Component/Language",
  component: OerebLanguage,
  tags: ["autodocs"]
};

export const Language = () => <OerebLanguage />;
