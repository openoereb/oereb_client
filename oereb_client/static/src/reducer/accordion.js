import {createSlice} from "@reduxjs/toolkit";

export const accordionSlice = createSlice({
  name: 'accordion',
  initialState: {
    category: null,
    topic: null,
    viewServices: []
  },
  reducers: {
    setActiveCategory: (state, action) => {
      state.category = action.payload;
    },
    setActiveTopic: (state, action) => {
      state.topic = action.payload;
    },
    setViewServices: (state, action) => {
      state.viewServices = action.payload;
    }
  }
});

export const {setActiveCategory, setActiveTopic, setViewServices} = accordionSlice.actions;

export default accordionSlice.reducer;
