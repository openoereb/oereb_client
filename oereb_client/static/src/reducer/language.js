import {createSlice} from "@reduxjs/toolkit";

import i18n from "../i18n";

export const languageSlice = createSlice({
  name: 'language',
  initialState: {
    current: null,
    default: null,
    available: []
  },
  reducers: {
    initLanguages: (state, action) => {
      state.default = action.payload.default;
      state.available = action.payload.available;
      const query = new URLSearchParams(window.location.search);
      if (query.has('lang') && state.available.indexOf(query.get('lang')) > -1) {
        state.current = query.get('lang');
      }
      else {
        state.current = state.default;
      }
      i18n.changeLanguage(state.current);
      query.set('lang', state.current);
      window.history.pushState(null, null, '?' + query.toString());
    },
    setLanguage: (state, action) => {
      if (state.available.indexOf(action.payload) > -1) {
        state.current = action.payload;
        i18n.changeLanguage(state.current);
        const query = new URLSearchParams(window.location.search);
        query.set('lang', state.current);
        window.history.pushState(null, null, '?' + query.toString());
      }
    }
  }
});

export const {initLanguages, setLanguage} = languageSlice.actions;

export default languageSlice.reducer;
