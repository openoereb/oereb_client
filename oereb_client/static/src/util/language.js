import {isArray} from "lodash";

export const getLocalizedText = function (multilingualText, language, defaultLanguage) {
  if (isArray(multilingualText) && multilingualText.length > 0) {
    let defaultValue = null;
    for (let i = 0; i < multilingualText.length; i++) {
      if (multilingualText[i]["Language"] === language) {
        return multilingualText[i]["Text"];
      }
      else if (multilingualText[i]["Language"] === defaultLanguage) {
        defaultValue = multilingualText[i]["Text"];
      }
    }
    if (defaultValue) {
      return defaultValue;
    }
    return multilingualText[0]["Text"];
  }
  return null;
};

export const getLocalizedUrl = function (multilingualUrl, language, defaultLanguage) {
  if (isArray(multilingualUrl) && multilingualUrl.length > 0) {
    let defaultValue = null;
    for (let i = 0; i < multilingualUrl.length; i++) {
      if (multilingualUrl[i]["Language"] === language) {
        return multilingualUrl[i]["URL"];
      }
      else if (multilingualUrl[i]["Language"] === defaultLanguage) {
        defaultValue = multilingualUrl[i]["URL"];
      }
    }
    if (defaultValue) {
      return defaultValue;
    }
    return multilingualUrl[0]["URL"];
  }
  return null;
};
