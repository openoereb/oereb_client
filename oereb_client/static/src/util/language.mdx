import {Meta} from '@storybook/addon-docs/blocks';

<Meta title="API Reference/Util/Language" />

# getLocalizedText

Accepts a multilingual array containing localized texts for certain languages,
the requested language and the default language. The localized text of the
requested language is returned as result. If the requested language is not
available in the array, the default language is used as fallback.

```js
getLocalizedText(multilingualText, language, defaultLanguage);
```

### Parameters

| Param | Type | Description |
| --- | --- | --- |
| multilingualText | <code>Array</code> | The array containing the localized texts. |
| language | <code>String</code> | The requested language. |
| defaultLanguage | <code>String</code> | The default language used as fallback. |

### Returns

<code>String</code> The localized text.

# getLocalizedUrl

Accepts a multilingual array containing localized URLs for certain languages,
the requested language and the default language. The localized URL of the
requested language is returned as result. If the requested language is not
available in the array, the default language is used as fallback.

```js
getLocalizedUrl(multilingualUrl, language, defaultLanguage);
```

### Parameters

| Param | Type | Description |
| --- | --- | --- |
| multilingualUrl | <code>Array</code> | The array containing the localized URLs. |
| language | <code>String</code> | The requested language. |
| defaultLanguage | <code>String</code> | The default language used as fallback. |

### Returns

<code>String</code> The localized URL.
