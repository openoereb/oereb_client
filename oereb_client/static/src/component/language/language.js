import PropTypes from 'prop-types';
import React from "react";
import {withNamespaces} from 'react-i18next';
import {useDispatch, useSelector} from "react-redux";

import {setLanguage} from "../../reducer/language";

const OerebLanguage = function({t}) {
  const dispatch = useDispatch();
  const language = useSelector((state) => state.language);
  const currentLanguage = language.current;
  const availableLanguages = language.available;

  const selectLanguage = function(lang) {
    dispatch(setLanguage(lang));
  };

  const languages = availableLanguages.map((language, key) =>
    <li key={key}>
      <button className="dropdown-item" onClick={selectLanguage.bind(this, language)}>
        {language.toUpperCase()}
      </button>
    </li>
  );

  return (
    <div>
      <button className="btn btn-outline-secondary dropdown-toggle"
        type="button"
        data-bs-toggle="dropdown"
        aria-expanded="false"
        title={t('menu.language.title')}>
        {currentLanguage.toUpperCase()}
      </button>
      <ul className="dropdown-menu">
        {languages}
      </ul>
    </div>
  );
};

OerebLanguage.propTypes = {
  t: PropTypes.func.isRequired
};

export default withNamespaces()(OerebLanguage);
