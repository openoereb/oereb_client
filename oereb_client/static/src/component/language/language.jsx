import React from "react";
import {useTranslation} from "react-i18next";
import {useDispatch, useSelector} from "react-redux";

import {hideExtract, loadExtract, showError, showExtract} from "../../reducer/extract";
import {updateHistory} from "../../reducer/history";
import {setLanguage} from "../../reducer/language";
import {queryExtractById} from "../../request/extract";
import {TooManyRequestsError} from "../../util/error";
import {showWarning} from "../../reducer/message";

/**
 * This component provides the language selector.
 */
const OerebLanguage = function() {
  const {t} = useTranslation();
  const dispatch = useDispatch();
  const config = useSelector((state) => state.config).config;
  const serviceUrl = config.service_url;
  const language = useSelector((state) => state.language);
  const currentLanguage = language.current;
  const availableLanguages = language.available;
  const extract = useSelector((state) => state.extract);

  const selectLanguage = function(lang) {
    if (lang !== currentLanguage) {
      dispatch(setLanguage(lang));
      if (extract.visible) {
        dispatch(loadExtract({
          egrid: extract.egrid,
          identdn: extract.identdn,
          number: extract.number,
          zoom: false
        }));
        queryExtractById(
          serviceUrl,
          extract.egrid,
          extract.identdn,
          extract.number,
          config.extract_json_timeout,
          lang
        )
          .then((data) => {
            dispatch(showExtract(data));
            dispatch(updateHistory(data));
          })
          .catch((error) => {
            if (error instanceof TooManyRequestsError) {
              dispatch(hideExtract());
              dispatch(showWarning(t('extract.error.too_many_requests'), true));
            }
            else {
              dispatch(showError());
            }
          });
      }
    }
  };

  const languages = availableLanguages.map((language, key) =>
    <li key={key}>
      <button className="dropdown-item" onClick={selectLanguage.bind(this, language)}>
        {language.toUpperCase()}
      </button>
    </li>
  );

  return (
    <div className="oereb-client-language-selector">
      <button className="btn btn-outline-secondary dropdown-toggle"
        type="button"
        data-bs-toggle="dropdown"
        aria-expanded="false"
        title={t("menu.language.title")}>
        {currentLanguage.toUpperCase()}
      </button>
      <ul className="dropdown-menu">
        {languages}
      </ul>
    </div>
  );
};

export default OerebLanguage;
