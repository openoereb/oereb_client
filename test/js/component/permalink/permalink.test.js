import {mount} from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";
import {render} from "react-dom";
import {act} from "react-dom/test-utils";
import {Provider} from "react-redux";

import OerebPermalink from "../../../../oereb_client/static/src/component/permalink/permalink";
import {initLanguages} from "../../../../oereb_client/static/src/reducer/language";
import MainStore from "../../../../oereb_client/static/src/store/main";

describe('permalink component', () => {

  let component;

  const modal =
    <div className="modal fade" id="permalinkModal" tabIndex="-1" aria-hidden="true">
      <div className="modal-dialog modal-lg">
        <div className="modal-content">
          <div className="modal-header">
            <h5 className="modal-title">Permalink</h5>
            <button type="button"
              className="btn-close"
              data-bs-dismiss="modal"
              aria-label="Close">
            </button>
          </div>
          <div className="modal-body">
            <input className="form-control" />
          </div>
        </div>
      </div>
    </div>;
  beforeEach(() => {
    act(() => {
      MainStore.dispatch(initLanguages({
        default: 'de',
        available: ['de']
      }));
    });
    const el = document.createElement('div');
    document.body.append(el);
    render(modal, el);
    component = mount(
      <Provider store={MainStore}>
        <OerebPermalink />
      </Provider>
    );
  });

  it('should render permalink modal', () => {
    expect(toJson(component)).toMatchSnapshot();
    component.find('button').simulate('click');
    expect(toJson(component)).toMatchSnapshot();
  });

});