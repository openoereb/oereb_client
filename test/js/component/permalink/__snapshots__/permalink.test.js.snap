// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`permalink component should render permalink modal 1`] = `
<DocumentFragment>
  <button
    class="btn btn-outline-secondary oereb-client-permalink"
    title="extract.permalink.title"
    type="button"
  >
    <i
      class="bi bi-link-45deg"
    />
  </button>
</DocumentFragment>
`;

exports[`permalink component should render permalink modal 2`] = `
<DocumentFragment>
  <button
    class="btn btn-outline-secondary oereb-client-permalink"
    title="extract.permalink.title"
    type="button"
  >
    <i
      class="bi bi-link-45deg"
    />
  </button>
</DocumentFragment>
`;
